native/membrane_videocompositor/src/elixir_bridge/scene_validation.rs

```diff
@@ -0,0 +1,456 @@
+        for (name, _) in objects {
+            if names.contains(name) {
+                return Err(SceneParsingError::DuplicateNames(name.clone()));
+            }
+
+            names.insert(name);
+        }
+
+        for (_, object) in objects {
+            for name in object.mentioned_names() {
+                if !names.contains(name) {
+                    return Err(SceneParsingError::UndefinedName(name.clone()));
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn check_for_unused_objects(
+        objects: &[(ObjectName, Object)],
+        final_object_name: &ObjectName,
+    ) -> Result<(), SceneParsingError> {
+        let defined_names = objects.iter().map(|(name, _)| name).collect::<HashSet<_>>();
+
+        let mut used_names = objects
+            .iter()
+            .flat_map(|(_, object)| object.mentioned_names())
+            .collect::<HashSet<_>>();
+
+        if used_names.contains(final_object_name) {
+            return Err(SceneParsingError::CycleDetected);
+        } else {
+            used_names.insert(final_object_name);
+        }
+
+        let unused_names = defined_names.difference(&used_names).collect::<Vec<_>>();
+
+        if !unused_names.is_empty() {
+            return Err(SceneParsingError::UnusedObject((*unused_names[0]).clone()));
+        }
+
+        Ok(())
+    }
+
+    /// returns true if a cycle exists in the scene graph
+    fn contains_cycle(
+        objects: &HashMap<&ObjectName, &Object>,
+        final_object: &ObjectName,
+    ) -> Result<(), SceneParsingError> {
+        enum NodeState {
+            BeingVisited,
+            Visited,
+        }
+
+        fn visit(
+            name: &ObjectName,
+            objects: &HashMap<&ObjectName, &Object>,
+            visited: &mut HashMap<ObjectName, NodeState>,
+        ) -> Result<(), SceneParsingError> {
+            match visited.get(name) {
+                Some(NodeState::BeingVisited) => return Err(SceneParsingError::CycleDetected),
+                Some(NodeState::Visited) => return Ok(()),
+                _ => {}
+            }
+
+            visited.insert(name.clone(), NodeState::BeingVisited);
+
+            for child in objects[name].previous_names() {
+                visit(child, objects, visited)?;
+            }
+
+            visited.insert(name.clone(), NodeState::Visited);
+
+            Ok(())
+        }
+
+        let mut visited = HashMap::new();
+
+        visit(final_object, objects, &mut visited)
+    }
+
+    pub fn validate(&self) -> Result<(), SceneParsingError> {
+        Self::check_for_duplicate_pad_refs(&self.objects)?;
+
+        Self::check_for_duplicate_or_undefined_names(&self.objects, &self.output)?;
+
+        Self::check_for_unused_objects(&self.objects, &self.output)?;
+
+        let objects = self
+            .objects
+            .iter()
+            .map(|(name, object)| (name, object))
+            .collect::<HashMap<_, _>>();
```

Comment:
[NIT] This collecting procedure is on a different level of abstraction compared to the rest of the body of this function.
I suggest extracting it to the separate helper function.